tests/ssl-opt.sh

```diff
@@ -3636,6 +3636,289 @@ run_test    "Large packet TLS 1.2 AEAD shorter tag" \
             -c "16384 bytes written in 1 fragments" \
             -s "Read from client: 16384 bytes read"
 
+# Tests of asynchronous private key support in SSL
+
+requires_config_enabled MBEDTLS_SSL_ASYNC_PRIVATE_C
+run_test    "SSL async private: sign, delay=0" \
+            "$P_SRV \
+             async_operations=s async_private_delay1=0 async_private_delay2=0" \
+            "$P_CLI" \
+            0 \
+            -s "Async sign callback: using key slot " \
```

Comment:
There's no race condition here. The code doesn't sign with multiple keys in parallel! It selects the first that has the right type considering the selected ciphersuite. Also the delay is in steps, not in time units, so it's deterministic. The slot number is an internal detail and the test script avoids relying on it when it isn't relevant.

Having different delays for the two keys was an unintended aspect of the design, due to the fact that the `async_private_delay$n` parameters control two things: whether to use the key (-1 vs ≥0) and what callback sequencing to exercise if the key is used (the non-negative value). Since only one key is used, only one callback sequencing parameter is used to the value should not be repeated on the command line. I plan to separate the delay parameter from the selection of keys to make asynchronous.

Reply:
This was a misunderstanding based on the strings being matched. If there is no race condition here, and its deterministic, can't we expect the exact key slot that will be used? Why leave out the slot number from the test string?